idx/cassandra/cassandra.go

```diff
@@ -236,46 +237,68 @@ func (c *CasIdx) Stop() {
+		return archive
+	}
 
-	if inMemory {
-		if existing.Partition == partition {
-			var oldest time.Time
-			if updateInterval > 0 {
-				oldest = time.Now().Add(-1 * updateInterval).Add(-1 * time.Duration(rand.Int63n(updateInterval.Nanoseconds()*int64(updateFuzzyness*100)/100)))
-			} else {
-				oldest = time.Now()
-			}
-			updateIdx = (existing.LastUpdate < oldest.Unix())
-		} else {
-			if updateCassIdx {
-				// the partition of the metric has changed. So we need to delete
-				// the current metricDef from cassandra.  We do this in a separate
-				// goroutine as we dont want to block waiting for the delete to succeed.
-				go func() {
-					if err := c.deleteDef(&existing); err != nil {
-						log.Error(3, err.Error())
-					}
-				}()
+	now := uint32(time.Now().Unix())
+
+	// Cassandra uses partition id asthe partitionin key, so an "update" that changes the partition for
+	// an existing metricDef will just create a new row in the table and wont remove the old row.
+	// So we need to explicitly delete the old entry.
+	if inMemory && existing.Partition != partition {
+		go func() {
+			if err := c.deleteDef(&existing); err != nil {
+				log.Error(3, err.Error())
 			}
-			updateIdx = true
-		}
-	} else {
-		updateIdx = true
-		stat = statAddDuration
+		}()
 	}
 
-	if updateIdx {
-		archive := c.MemoryIdx.AddOrUpdate(data, partition)
-		if updateCassIdx {
-			log.Debug("cassandra-idx updating def in index.")
-			c.writeQueue <- writeReq{recvTime: time.Now(), def: &archive.MetricDefinition}
-		}
+	// check if we need to save to cassandra.
+	if archive.LastSave >= (now - updateInterval32) {
 		stat.Value(time.Since(pre))
 		return archive
 	}
-	return existing
+
+	// This is just a safety precaution to prevent corrupt index entries.
+	// This ensures that the index entry always contains the correct metricDefinition data.
+	if inMemory {
+		archive.MetricDefinition = *schema.MetricDefinitionFromMetricData(data)
+		archive.MetricDefinition.Partition = partition
+	}
+
+	// if the entry has not been saved for 1.5x updateInterval
+	// then perform a blocking save. (bit shifting to the right 1 bit, divides by 2)
+	if archive.LastSave < (now - updateInterval32 - (updateInterval32 >> 1)) {
+		log.Debug("cassandra-idx updating def in index.")
+		c.writeQueue <- writeReq{recvTime: time.Now(), def: &archive.MetricDefinition}
+		archive.LastSave = now
+		c.MemoryIdx.Update(archive)
+	} else {
+		// perform a non-blocking write to the writeQueue. If the queue is full, then
+		// this will fail and we wont update the LastSave timestamp. The next time
+		// the metric is seen, the previous lastSave timestamp will still be in place and so
+		// we will try and save again.  This will continue until we are successful or the
+		// lastSave timestamp become more then 1.5 x UpdateInterval, in which case we will
+		// do a blocking write to the queue.
+		select {
+		case c.writeQueue <- writeReq{recvTime: time.Now(), def: &archive.MetricDefinition}:
```

Comment:
faster? yes.  But as we use this for measuring how long items are spending in the queue `time.Unix(now, 0)` does not provide the required precision